### DISCUSSION TITLE
Can I get the last run command from a kitten?

### DISCUSSION THREAD

**Discussion by groves** [kovidgoyal/kitty#5414] [Q&A] [answered]
I'd like to know which command the user executed in the shell to be able to run command-specific hinters on the command output.  For example, if the last command started with rg, I'd hint ripgrep style output whereas if it started with cargo, I'd look for output like the rust compiler produces.

I was thinking I could do that with the OSC 133 B command from https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md#commands, but it doesn't seem like kitty passes the B mark through ansi formatted input for kittens. I get 133 A and 133 D, but not B.

My current fallback thinking is to emit an OSC 8 url with a scheme specific to this and use that as my marker e.g. `echo -en '\e]8;;prompt://end\e\\>\e]8;;\e\\'` if `>` is my end of prompt marker. 

Does that seem like a reasonable way to do it or is there a better approach?

**Comment by kovidgoyal** [comment, marked as answer]
If I were you I would instead just write wrappers for the commands in question, that use OSC 8 hyperlinks directly, see for example, the hyperlinked_grep kitten. The you can just use the hints kitten in hyperlink mode.

**Reply by groves** [reply]
Fair enough. I don't like wrappers as much since you have to change how you're invoking things, but they totally work.

**Comment by kovidgoyal** [comment]
Implementing B robustly for all shells is not worth the effort, at least
for me. A should be enough for your purposes, since you know what the
prompt looks like your kitten can simply extract the command from the
line on which the prompt starts.

**Reply by groves** [reply]
> Implementing B robustly for all shells is not worth the effort

Yeah, implementing it in someone else's shell sounds really hard to me. I don't need that though, just that if a shell emits it that Kitty will pass it through.

> A should be enough for your purposes, since you know what the
prompt looks like

By that logic, I don't really need A, either. Having The prompt markers means I can write a kitten that works for anyone who's configured their shell, not just me. Having support for all of A, B, and D and tools that use them make it more likely that shell and prompt implementors will start using them.

Would you be opposed to adding B tracking and emission if I submitted a PR for it?

**Comment by kovidgoyal** [comment]
On Thu, Aug 25, 2022 at 03:23:11AM -0700, Charlie Groves wrote:
> > Implementing B robustly for all shells is not worth the effort
> 
> Yeah, implementing it in someone else's shell sounds really hard to me. I don't need that though, just that if a shell emits it that Kitty will pass it through.
> 
> > A should be enough for your purposes, since you know what the
> prompt looks like
> 
> By that logic, I don't really need A, either. Having The prompt markers means I can write a kitten that works for anyone who's configured their shell, not just me. Having support for all of A, B, and D and tools that use them make it more likely that shell and prompt implementors will start using them.
> 
> Would you be opposed to adding B tracking and emission if I submitted a PR for it?

Depends on the implementation details. This requires tracking per cell
whereas the others are per line. So it will likely not be low overhead.

As I said in my other post, if I were you I would implement this as a
wrapper program hyperlinkify. Then your putative other users could
simply add

alias rg="hyperlinkify rg"
alias whatever="hyperlinkify whatever"

Doesn't need any special support from the terminal emulator (other than
hyperlinks) and so works everywhere.

**Reply by groves** [reply]
> Depends on the implementation details. This requires tracking per cell whereas the others are per line. So it will likely not be low overhead.

I was thinking I could make the per-cell hyperlink id field more general to do this. We could reserve some portion of the id space for known builtin marks and emit those marks when we hit that id. https://github.com/kovidgoyal/kitty/blob/master/kitty/line.c#L354 makes it look like that wouldn't be too awful.

**Comment by kovidgoyal** [comment]
On Thu, Aug 25, 2022 at 04:24:40AM -0700, Charlie Groves wrote:
> > Depends on the implementation details. This requires tracking per cell whereas the others are per line. So it will likely not be low overhead.
> 
> I was thinking I could make the per-cell hyperlink id field more general to do this. We could reserve some portion of the id space for known builtin marks and emit those marks when we hit that id. https://github.com/kovidgoyal/kitty/blob/master/kitty/line.c#L354 makes it look like that wouldn't be too awful.

The id space is only 16 bits, that would need to come down to 14bits.
Doesn't seem worth it to me. If you really want to do this then store it 
in ch rather than hyperlink_id. ch is 32bits of which 12 are wasted,
since the largest number in unicode is 20bits IIRC.
